rewrite-core/src/main/java/org/openrewrite/semver/XRange.java

```diff
@@ -46,7 +46,7 @@ public class XRange extends LatestRelease {
     public boolean isValid(@Nullable String currentVersion, String version) {
         // For X-range patterns, we need to check if the version matches the pattern
         // but we should not exclude it just because it's a pre-release version
-        if (!VersionComparator.checkVersion(version, getMetadataPattern(), false)) {
+        if (!VersionComparator.checkVersion(version, getMetadataPattern(), getMetadataPattern() == null)) {
```

Comment:
I think it would make sense to add a test case which demonstrates that a metadata pattern like `-beta.*` could be used to allow upgrades to a beta (non-release) version.

Reply:
So if you provide the `metadataPattern`, it's going to ensure any possible upgraded version also has that `metadataPattern`, even if there are higher release versions. Is that what we want to show?

i.e. You're at `3.4.0` and you tell it to upgrade to `3.5.x` with a metadata pattern of `-beta`. If our available versions for `3.5.x` are `3.5.2` and `3.5.1-beta`, it's going to take `3.5.1-beta` over `3.5.2` because of the `metadataPattern`.

Reply:
Good question. Would it then pick an older beta if the latest release doesn't have a beta? That wouldn't be what we want. Then we should instead just pass in `true` here.

```suggestion
        if (!VersionComparator.checkVersion(version, getMetadataPattern(), true)) {
```

Reply:
This is an area where Maven versioning clashes a bit with semver, unfortunately.